src/ORM/Hierarchy/Hierarchy.php

```diff
@@ -360,6 +381,127 @@ public function AllChildrenIncludingDeleted()
+     *   well to save future queries
+     * - Does not included deleted records, which could be done via Versioned::updateListToAlsoIncludeDeleted($children)
+     */
+    public function getAllChildren(): ArrayList
+    {
+        $owner = $this->getOwner();
+        $parentID = $owner->ID;
+        $baseClass = Hierarchy::getBaseHierarchyClass($owner);
+        Hierarchy::$all_children_child_ids_cache[$baseClass] ??= [];
+        Hierarchy::$all_children_obj_cache[$baseClass] ??= [];
+        if (!array_key_exists($parentID, Hierarchy::$all_children_obj_cache[$baseClass])) {
+            Hierarchy::$all_children_obj_cache[$baseClass][$parentID] = $owner;
+        }
+        $this->recursivelyCacheDescendants($baseClass, $parentID);
+        $childIDs = [];
+        if (array_key_exists($parentID, Hierarchy::$all_children_child_ids_cache[$baseClass])) {
+            $childIDs = Hierarchy::$all_children_child_ids_cache[$baseClass][$parentID];
+        }
+        $childNodes = [];
+        foreach ($childIDs as $childID) {
+            if (array_key_exists($childID, Hierarchy::$all_children_obj_cache[$baseClass])) {
+                $childNodes[] = Hierarchy::$all_children_obj_cache[$baseClass][$childID];
+            }
+        }
+        $children = ArrayList::create($childNodes);
+        $owner->extend('updateGetAllChildren', $includeDeleted, $children);
+        return $children;
+    }
+
+    /**
+     * Get the base class with the Hierarchy extension of a DataObject
+     * We need to get this instead of just getting the DataObjectSchema::baseDataClass() to ensure that we
+     * can run DB queries that filter on ParentID
+     */
+    public static function getBaseHierarchyClass(DataObject $obj): string
+    {
+        $baseHierarchyClass = $obj->ClassName;
+        $haveFoundTop = false;
+        do {
+            $parentDataClass = get_parent_class($baseHierarchyClass);
+            if (!$parentDataClass
+                || !$parentDataClass::has_extension(Hierarchy::class)
+                || !is_a($parentDataClass, DataObject::class, true)
+            ) {
+                $haveFoundTop = true;
+            } else {
+                $baseHierarchyClass = $parentDataClass;
+            }
+        } while (!$haveFoundTop);
+        return $baseHierarchyClass;
+    }
+
+    /**
+     * Inner recursive method used by getAllChildren
+     * This will use a hierarchical query method where it will initially fetched from the "second level down"
+     * where the ParentID equals the single DataObject at the "top level" of the hierarchy
+     * It will then fetch from the "third level down" in the hierarchy where the ParentID is of the ParentIDs in
+     * the "second level down". In will continue querying like this until the total number of cached objects has
+     * exceeded the configured value `node_threshold_total`, or all object have been fetched
```

Comment:
```suggestion
     * This recursively fetches children of the records in $idOrIDs until the total number of cached objects has
     * exceeded the configured value `node_threshold_total`, or all object have been fetched
```
No need to explain recursion in a method with "recursively" in its name

Reply:
It's not a regular recursion method where you'd just fetch nodes that match a single parentID, you instead have a "level" of parentIDs you fetch from which is more query efficient. I think it's worth pointing that out here

Reply:
As discussed online, the implementation details should be included as inline comments inside the method itself.
The PHPDoc should be agnostic to the inner workings - when calling the method you need to know what you pass in, and what you get in return. The fact this is cached is relevant, and the fact it recursively fetches each level is potentially relevant, but that can be described in one or two sentences.

Feel free to move the specific detail as comments into the method itself.

Reply:
Updated